### DISCUSSION TITLE
Extract Kernel mount/umount parts of the `cephcsi` binary?

### DISCUSSION THREAD

**Discussion by leseb** [ceph/ceph-csi#2950] [Q&A]
This is a follow-up to https://github.com/ceph/ceph-csi/discussions/2946 which has been answered. The clean takeaway (that I agree with) is the current `cephcsi` will never be statically compiled since we use `cgo` with `go-ceph`.

That said I'm still exploring other possibilities to get a statically compiled binary that would be responsible **only** for mounting and unmounting volumes using Kernel modules. 
In that sense, I'm wondering if it would be possible to extract the kernel mount/umount pieces from `cephcsi` and put that in another binary?

Thanks!

**Comment by BlaineEXE** [comment]
Will said kernel mounter be able to create mounts in the network namespace of the long-lived pod that is running the csi binary?

**Reply by leseb** [reply]
The Kernel mounter becomes the long-lived pod in that scenario.

**Comment by Madhu-1** [comment]
Do you mean the nodeStage and NodeUnstage API which is responsible for rbd mapping and unmapping? If yes, we still use go-ceph API to talk to ceph cluster except `rbd map` command.

**Reply by leseb** [reply]
Since both RBD and CephFS are impacted by this bug yes. Essentially I'm trying to see if we can extract:

* `rbd map`
* `mount -t ceph`

calls to a new daemon that we can be statically compiled.

EDIT: STATICALLY COMPILED OF COURSE.

**Reply by nixpanic** [reply]
I guess that can be done, have a UNIX-domain-socket that receives mount/attach/cryptsetup/... calls. That way `cephcsi` would not call these commands anymore.

It does add complexity, maintenance and a little performance overhead. Instead of a single `cephcsi` component, features might require implementing them in both components. And both components would need a restart too?

**Reply by leseb** [reply]
The mounter/umounter will never get restarted. The binary will be swapped on updates but the container will always remain alive (until the node reboots). That way we can ensure IO availability through upgrades of Ceph-CSI.

**Reply by BlaineEXE** [reply]
> Since both RBD and CephFS are impacted by this bug yes. Essentially I'm trying to see if we can extract:
> 
> * `rbd map`
> * `mount -t ceph`
> 
> calls to a new daemon that we can dynamically compile.

Do you mean statically compile? Perhaps I'm misunderstanding something, but my understanding was that we need a static binary in the always-alive container so that the container won't need to be updated with dependencies when a new binary is swapped in.

**Reply by leseb** [reply]
Of course, I meant statically compiled 🤦🏻 !

**Reply by Madhu-1** [reply]
@leseb is this something needed before the 3.6.0 release? will discuss today with the team about this one.

**Reply by nixpanic** [reply]
I still do not see how we can statically link an executable with all the Ceph components/libraries included. Something needs to call the `rbd` command, `mount.cephfs` and potentially `rbd-nbd` and `fuse-cephfs`. It would be a lot of work to cover all the options and parameters of all these executables into a single binary. And that all requires static linking, which needs to be made available by the Ceph project (and -devel packages). Linux distributions like Fedora and CentOS do not provide statically linkable libraries, so that would need to be addressed as well.

**Reply by leseb** [reply]
`mount.cephfs` does not strictly require any packages as far as I know. All the mount options needs to be passed to `mount`, we just need the kernel module to be enabled. For `rbd map`, we can use the `/sys/bus/rbd/add` interface to map and `/sys/bus/rbd/remove` to unmap. Again, if we can do so, we don't need any Ceph packages, just the kernel module. @idryomov do you see any issues with that?

**Reply by nixpanic** [reply]
Of course we can re-implement everything in a single executable. We could even use the syscall `mount()` and not exec `mount` at all. It is mostly a question if we should re-implement everything, and what we need to do when we have to fix bugs. I do not think such a static linked executable should be part of Ceph-CSI, as it touches many internals to Ceph. It would be better for the Ceph project to provide and maintain such an executable.

**Reply by leseb** [reply]
Sorry if I wasn't clear but the request is not to create a new binary for ceph-csi but more to decouple the mount logic. Regardless of where the binary lives we still need to call it from ceph-csi in the newly decoupled component. This binary would need to run autonomously.

**Reply by nixpanic** [reply]
Yes, that sounds technically possible to me. As long as there is some UNIX-Domain-Socket communication possible (HostPath?), ideally over gRPC, it would work. Ceph-CSI can be adapted to call the mounting, (d)encrypting, mapping and other commands through a different interface than the local executables.

Still, a complete static linked executable with all the tools and functionalities combined seems like a maintenance headache to me.

**Reply by idryomov** [reply]
The `mount`/`fsmount` syscalls and `/sys/bus/rbd/{add*,remove*}` are stable interfaces for which pretty much infinite backwards compatibility is guaranteed so anyone is free to cut out `rbd map` and `mount.ceph` helpers and use the kernel interfaces directly -- subject to "re-implement everything", "have to fix bugs" and other good concerns raised by @nixpanic of course.

But I'm not sure I understand where this is coming from.  It seems like the underlying issue is still the restart/upgrade of Ceph CSI containers that end up owning the network namespace.  I thought that was solved with a clever use of bridging suggested by the OCP networking folks.  Did that solution not pan out?

**Reply by leseb** [reply]
> The `mount`/`fsmount` syscalls and `/sys/bus/rbd/{add*,remove*}` are stable interfaces for which pretty much infinite backwards compatibility is guaranteed so anyone is free to cut out `rbd map` and `mount.ceph` helpers and use the kernel interfaces directly -- subject to "re-implement everything", "have to fix bugs" and other good concerns raised by @nixpanic of course.
> 

Thanks!

> But I'm not sure I understand where this is coming from. It seems like the underlying issue is still the restart/upgrade of Ceph CSI containers that end up owning the network namespace. I thought that was solved with a clever use of bridging suggested by the OCP networking folks. Did that solution not pan out?

This solution was considered for a long time until we realized the number of corner cases it had. I'm sure @BlaineEXE can elaborate a bit more but from the top of my head:

* the solution would work for the cri-o runtime only (not Docker)
* risk of orphan interfaces left on the host (however the IP is released from the IPAM) which would cause L2 connectivity issues
* Multus has different primary operating modes (macvlan, ipvlan, bridge, host-device), and “moving” the interface is not the same in every case
* moved interface is completely left out of the Multus's scope
* overall over-engineered solution, no real idempotency while moving the interface on the host

You can find the latest design proposal here https://github.com/rook/rook/pull/9903

**Reply by BlaineEXE** [reply]
Yes, that's a good summary from Seb. The bridging solution also might eliminate the possibility users have to use performance tweaking options for the multus interfaces without complicated caveats. 

Another key downside of the previously-proposed bridging solution is that we still force a solution where we break network isolation. In my estimation, network isolation is one of the largest benefits of using multus instead of host networking. The bridging solution effectively creates a network back door on all nodes.

The solution Seb is proposing here allows us to keep CSI kernel mounts stable through upgrades and still keep network isolation.

**Reply by idryomov** [reply]
Thanks for the explanation!  I have left my thoughts in the design proposal PR.
